frontend/views/containers/group-settings/GroupCreationModal.vue

```diff
@@ -105,6 +106,9 @@ export default {
         return
       }
 
+      if (this.ephemeral.isSubmitting) { return }
+      this.ephemeral.isSubmitting = true
```

Comment:
So this is a lot of the same code repeated everywhere these buttons need to be used. It reminds me of the same issue we had with the scoped-banners, and how we created the scoped banner component that is self-removing and therefore doesn't need any new variables or logic added to the parent component.

If we had a custom button component, it could automatically check itself to see if it's in the submitting state, by accepting a submit handler and waiting for the promise to finish. If it's clicked while the promise hasn't finished, it does nothing. This would remove the need for creating and checking this `ephemeral.isSubmitting` variable.

Do you think this is a component you could create? You could pass down all the other attributes that are passed to it to the `button` using [`v-bind="$attrs"`](https://vuejs.org/v2/api/#vm-attrs) and maybe [`v-on="$listeners"`](https://vuejs.org/v2/api/#vm-listeners) (although you'd have to be careful with that one, since we want to catch and wrap the `@click` submit handler ourselves for this component to do its self-contained magic).

Reply:
> Do you think this is a component you could create? 

When I joined Group Income this was one of the very first things I pointed about buttons. Not being a component. GI is the only "modern framework based" project I ever work on where a button isn't a Vue component.

That way we could just pass `isLoading` prop and it would handle [data-loading] correctly everywhere. The only ugly part is we would need to use `L` instead of `i18n` when using the component

```pug
button(:isLoading='isSubmitting' @click={submit})
  | {{ L('Submit') }}
```


However, we'd still need to add the verification on the submit handler (in the parent)  because the form may be submitted through another way (example: press Enter button) 

So... Your suggestion would work 90% of the places, but not everywhere (ex: on Login we can use Enter to login).

Reply:
> That way we could just pass `isLoading` prop 

My suggestion is to pass only the submit handler, e.g. `@click='submit'`. `isSubmitting` / `isLoading` doesn't need to be passed I don't think. Again the point is to avoid having to create these variables. The button will call `submit` and will change its state based on when the promise resolves or fails.

> However, we'd still need to add the verification on the submit handler (in the parent) because the form may be submitted through another way (example: press Enter button)

There's still gotta be a way to avoid introducing `isSubmitting`. For example, you could tag the button instance as a `ref` and then include a method on the button component that is called by the Enter handler.

e.g. something like:

```
@keyup.enter='$refs.submitButton.submit()'
```

Reply:
Hum... Using `$refs` could be an option. @pieer what do you think?

Reply:
So pressing 'enter' would call the button action which handle the loading state and call the parent submit function... That works for me.

Reply:
Exactly. I'll work on it then! 👩‍💻

Reply:
> Hum... Using $refs could be an option. @pieer what do you think?

Just dropping by a comment:

Using a ref would be considered bad practice due to the fact that if the internal API of the button changes at some point, the function that you are trying to call could not exist, hence creating a bug. So IF you were to go for a ref solution, you really need to be aware of changes.

Reply:
OK, I think I get it, the `$listeners` refers only to the event handlers registered on this instance of the component.

However, @sandrina-p, the use-case for `$refs` was when `@keyup.enter` was registered on another part of the form. In that case you'd need a `$refs` reference (as shown in my earlier comment above), wouldn't you?

Reply:
No, we don't need `$refs`. When pressing Enter, `form(@submit.prevent='')` gets called. And when a form is submitted, it automatically triggers its submit element action. In this case, `button-submit` is the submit element thanks to `type="submit"`. So, `@click` gets called directly.

Reply:
Would this solve your problem https://codepen.io/BlitZZ96/pen/RwPLEXN
This example uses a custom directive that acts like the default vue `@click`, but it updates the element that uses it's HTML to a loader while the promise is unresolved.

Reply:
> This example uses a custom directive

@BlitZz96 Interesting solution! I think that approach might work too! I think @sandrina-p's approach also works well though, particularly since we want to make sure it works well with the `<enter>` key when a form is submitted.

By using `@click`, `$listeners`, `form(@submit.prevent='')`, @sandrina-p has ensured that the button will behave correctly and in the same way regardless of whether the form was submitted by clicking on the button, or because the user pressed enter.

Reply:
One thing about directives: they don't have the ability to access and update component state right? We need to make sure that this button does nothing when it's clicked while it's in the loading state. So if the directive isn't able to toggle component state, then the component-based approach is the only thing that works.

Reply:
@taoeffect 
Okey, then these simple changes would do the job
https://codepen.io/BlitZZ96/pen/WNvZmZW

If simplicity is what you are after.. This is definitly simpler than using `@click`, `$listen` and `form(@submit.prevent='')`

Reply:
> We need to make sure that this button does nothing when it's clicked while it's in the loading state

This could be achieved by doing `el.disabled = true;` before the callback and false in `.then()`.

Well, I want you guys to do what is simpler for you, but ofc, both solutions has its pros and cons. One pro with this one is that you do not need a "loading" state in the parent, because the promise does that for you. Other plus is that you do not need to add a loader everywhere. A con might be that It needs some polishing in order to work IDENTICAL to @click (with the event modifiers .prevent, .native, .stop etc...)

Reply:
@BlitZz96 Think you might've missed my comment above about the state being used to prevent double submissions. In your example, pressing the enter key twice in rapid succession will result in the button permanently in a loading state, and `handleSubmit` getting called twice.

Reply:
(Deleted my other comment, we posted at the same time.)

> This could be achieved by doing `el.disabled = true;` before the callback and false in `.then()`. or something similar

Ah, yes that looks like it could work!

> both solutions has its pros and cons.

Yeah, I like both solutions! Will be curious to hear what @sandrina-p thinks when she's back tomorrow.

If we go with @sandrina-p's approach, we should still bookmark yours, maybe via a link to these comments in the source. Thank you so much @BlitZz96 for sharing this! Always amazing to see so many different ways to solve a problem with Vue.js!

Reply:
No problem @taoeffect! This was probably more fun for me than you guys 😂 Thanks for the interesting problem.

Reply:
@BlitZz96 I learned something new thanks to you! You are always welcome to drop by! ^_^

Reply:
Hi @BlitZz96  👋 , thanks for your help

> Using a ref would be considered bad practice due to the fact that if the internal API of the button changes at some point.

I used to think the same. But then I realized that using `props` suffers from the same problem. The moment you change/delete a prop, the component can break. At the moment we only have one component that is used through `$refs` and it's documented to prevent any internal API accidental change.  You can check `components/banners/BannerScoped.vue`.

Now, back to the loading state. Using a `directive` as you proposed is an interesting solution! However, I find the `directive` not as efficient as the `ButtonSubmit` component because:
- It contains additional logic to handle the form using `el.querySelector`. `ButtonSubmit` doesn't need it.
- We need to perform more DOM changes than just adding a spinner. One of them is to change `disabled`. However, when it's not loading (a.k.a `false`), it requires some conditional logic based on the context. For example, at `LoginForm` the `disabled` status also depends if the form is valid or not. This would be the API for each approach:

```pug
    button(
      v-save='{ handler: login, disabled: "$v.form.$invalid" }'
    ) Login

    button-submit(
      @click='login'
      :disabled='$v.form.$invalid'
    ) Login
```

Using a solution based on component + props is cleaner to use and easier to write thanks to its declarative way that Vue offers us, which is not possible when using `directive`.

TLDR: I prefer using the component `ButtonSubmit` over the directive `v-save` because it's more flexible and cleaner.